src/prefect/server/services/task_run_recorder.py

```diff
@@ -1,14 +1,179 @@
+):
+    await session.execute(
+        db.insert(db.TaskRunState)
+        .values(
+            created=pendulum.now("UTC"),
+            task_run_id=task_run.id,
+            **task_run.state.model_dump(),
+        )
+        .on_conflict_do_nothing(
+            index_elements=[
+                "id",
+            ]
+        )
+    )
+
+
+@db_injector
+async def _update_task_run_with_state(
+    db: PrefectDBInterface,
+    session: AsyncSession,
+    task_run: TaskRun,
+    denormalized_state_attributes: Dict[str, Any],
+):
+    await session.execute(
+        sa.update(db.TaskRun)
+        .where(
+            db.TaskRun.id == task_run.id,
+            sa.or_(
+                db.TaskRun.state_timestamp.is_(None),
+                db.TaskRun.state_timestamp < task_run.state.timestamp,
+            ),
+        )
+        .values(**denormalized_state_attributes)
+    )
+
+
+def task_run_from_event(event: ReceivedEvent) -> TaskRun:
+    task_run_id = event.resource.prefect_object_id("prefect.task-run")
+
+    flow_run_id: Optional[UUID] = None
+    if flow_run_resource := event.resource_in_role.get("flow-run"):
+        flow_run_id = flow_run_resource.prefect_object_id("prefect.flow-run")
+
+    state: State = State.model_validate(
+        {
+            "id": event.id,
+            "timestamp": event.occurred,
+            **event.payload["validated_state"],
+        }
+    )
+    state.state_details.task_run_id = task_run_id
+    state.state_details.flow_run_id = flow_run_id
+
+    return TaskRun.model_validate(
+        {
+            "id": task_run_id,
+            "flow_run_id": flow_run_id,
+            "state_id": state.id,
+            "state": state,
+            **event.payload["task_run"],
+        }
+    )
+
+
+async def record_task_run_event(event: ReceivedEvent, depth: int = 0):
+    db = provide_database_interface()
+
+    async with AsyncExitStack() as stack:
+        await stack.enter_async_context(
+            (
+                causal_ordering().preceding_event_confirmed(
+                    record_task_run_event, event, depth=depth
+                )
+            )
+        )
+
+        task_run = task_run_from_event(event)
+
+        task_run_attributes = task_run.model_dump_for_orm(
+            exclude={
+                "state_id",
+                "state",
+                "created",
+                "estimated_run_time",
+                "estimated_start_time_delta",
+            },
+            exclude_unset=True,
+        )
+
+        assert task_run.state
+
+        denormalized_state_attributes = {
+            "state_id": task_run.state.id,
+            "state_type": task_run.state.type,
+            "state_name": task_run.state.name,
+            "state_timestamp": task_run.state.timestamp,
+        }
+        session = await stack.enter_async_context(db.session_context())
+
+        await _insert_task_run(session, task_run, task_run_attributes)
```

Comment:
Unfortunately due to FK's we need to insert the task run,  _then_ insert the task run state, _then_ go back and update the task run with the denormalized state attributes

Reply:
Good question, that might make the lock a little more aggressive?  I'm thinking about the SQLite implementation hitting `database is locked`.  Since this should be the only thing writing task runs, I think we're good with just the transaction